In [ ]:
import pandas as pd

# Load feature and target datasets
X = pd.read_csv("derm_data.csv")
y = pd.read_csv("derm_target.csv").squeeze() # convert dataframe to 1D array to better fit SVC models

In [ ]:
X.head()

,erythema,scaling,definite-borders,itching,koebner phenomenon,polygonal papules,follicular papules,oral-mucosal involvement,knee elbow involvement,scalp involvement,...,focal hypergranulosis,disappearance of the granular layer,vacuolisation and damage of the basal layer,spongiosis,saw-tooth appearance of retes,follicular horn plug,perifollicular parakeratosis,inflammatory monoluclear infiltrate,band-like infiltrate,age
0,2,2,0,3,0,0,0,0,1,0,...,0,0,0,3,0,0,0,1,0,55.0
1,3,3,3,2,1,0,0,0,1,1,...,0,0,0,0,0,0,0,1,0,8.0
2,2,1,2,3,1,3,0,3,0,0,...,2,0,2,3,2,0,0,2,3,26.0
3,2,2,2,0,0,0,0,0,3,2,...,0,3,0,0,0,0,0,3,0,40.0
4,2,3,2,2,2,2,0,2,0,0,...,2,2,3,2,3,0,0,2,3,45.0


In [ ]:
y.head()

,class
0,2
1,1
2,3
3,1
4,3


In [ ]:
X.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 366 entries, 0 to 365
Data columns (total 34 columns):
 #   Column                                       Non-Null Count  Dtype  
---  ------                                       --------------  -----  
 0   erythema                                     366 non-null    int64  
 1   scaling                                      366 non-null    int64  
 2   definite-borders                             366 non-null    int64  
 3   itching                                      366 non-null    int64  
 4   koebner phenomenon                           366 non-null    int64  
 5   polygonal papules                            366 non-null    int64  
 6   follicular papules                           366 non-null    int64  
 7   oral-mucosal involvement                     366 non-null    int64  
 8   knee elbow involvement                       366 non-null    int64  
 9   scalp involvement                            366 non-null    int64  
 10  fa

In [ ]:
y.info()

<class 'pandas.core.series.Series'>
RangeIndex: 366 entries, 0 to 365
Series name: class
Non-Null Count  Dtype
--------------  -----
366 non-null    int64
dtypes: int64(1)
memory usage: 3.0 KB


.info() showed that all varibles are non-null other than 8 null age values, consistant with the information provided at the dataset sorce website.

In [ ]:
# Fill missing age values with the median to preserve distribution while avoiding NaNs
X['age'].fillna(X['age'].median(), inplace=True)

/tmp/ipykernel_3135/3336523914.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  X['age'].fillna(X['age'].median(), inplace=True)


In [ ]:
from sklearn.preprocessing import StandardScaler

# Scale all features for better compatibility with SVM models
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)
# Keep as a Dataframe
X_scaled = pd.DataFrame(X_scaled, columns=X.columns)

In [ ]:
# Preview data after scaling
print(X_scaled.head()) # scaled
print(y.head()) # unchanged

   erythema   scaling  definite-borders   itching  koebner phenomenon  \
0 -0.102895  0.292503         -1.709375  1.437334           -0.699049   
1  1.403482  1.719916          1.600843  0.557628            0.403761   
2 -0.102895 -1.134911          0.497437  1.437334            0.403761   
3 -0.102895  0.292503          0.497437 -1.201784           -0.699049   
4 -0.102895  1.719916          0.497437  0.557628            1.506571   

   polygonal papules  follicular papules  oral-mucosal involvement  \
0          -0.468702           -0.292496                 -0.452637   
1          -0.468702           -0.292496                 -0.452637   
2           2.669312           -0.292496                  3.148777   
3          -0.468702           -0.292496                 -0.452637   
4           1.623308           -0.292496                  1.948306   

   knee elbow involvement  scalp involvement  ...  focal hypergranulosis  \
0                0.392453          -0.574000  ...              -

In [ ]:
# Verify Mean of 0 and Std of 1
X_scaled.describe()

,erythema,scaling,definite-borders,itching,koebner phenomenon,polygonal papules,follicular papules,oral-mucosal involvement,knee elbow involvement,scalp involvement,...,focal hypergranulosis,disappearance of the granular layer,vacuolisation and damage of the basal layer,spongiosis,saw-tooth appearance of retes,follicular horn plug,perifollicular parakeratosis,inflammatory monoluclear infiltrate,band-like infiltrate,age
count,3.660000e+02,3.660000e+02,3.660000e+02,3.660000e+02,3.660000e+02,3.660000e+02,3.660000e+02,3.660000e+02,3.660000e+02,3.660000e+02,...,3.660000e+02,3.660000e+02,3.660000e+02,3.660000e+02,3.660000e+02,3.660000e+02,3.660000e+02,3.660000e+02,3.660000e+02,3.660000e+02
mean,-2.620854e-16,-1.650168e-16,-3.882747e-17,3.882747e-17,-4.853434e-17,-4.853434e-17,8.372174e-17,5.338777e-17,-9.706868e-18,-2.912060e-17,...,4.853434e-18,-7.765494e-17,-1.213358e-17,1.043488e-16,2.184045e-17,-3.397404e-17,-4.368091e-17,3.882747e-17,4.368091e-17,-2.135511e-16
std,1.001369e+00,1.001369e+00,1.001369e+00,1.001369e+00,1.001369e+00,1.001369e+00,1.001369e+00,1.001369e+00,1.001369e+00,1.001369e+00,...,1.001369e+00,1.001369e+00,1.001369e+00,1.001369e+00,1.001369e+00,1.001369e+00,1.001369e+00,1.001369e+00,1.001369e+00,1.001369e+00
min,-3.115648e+00,-2.562324e+00,-1.709375e+00,-1.201784e+00,-6.990489e-01,-4.687015e-01,-2.924961e-01,-4.526367e-01,-6.262552e-01,-5.739998e-01,...,-4.638313e-01,-5.377701e-01,-4.785022e-01,-8.448778e-01,-4.757013e-01,-2.308165e-01,-2.351254e-01,-2.573550e+00,-5.022153e-01,-2.396101e+00
25%,-1.028946e-01,-1.134911e+00,-6.059689e-01,-1.201784e+00,-6.990489e-01,-4.687015e-01,-2.924961e-01,-4.526367e-01,-6.262552e-01,-5.739998e-01,...,-4.638313e-01,-5.377701e-01,-4.785022e-01,-8.448778e-01,-4.757013e-01,-2.308165e-01,-2.351254e-01,-1.194459e+00,-5.022153e-01,-7.444268e-01
50%,-1.028946e-01,2.925027e-01,4.974372e-01,-3.220782e-01,-6.990489e-01,-4.687015e-01,-2.924961e-01,-4.526367e-01,-6.262552e-01,-5.739998e-01,...,-4.638313e-01,-5.377701e-01,-4.785022e-01,-8.448778e-01,-4.757013e-01,-2.308165e-01,-2.351254e-01,1.846324e-01,-5.022153e-01,-8.375705e-02
75%,-1.028946e-01,2.925027e-01,4.974372e-01,5.576279e-01,4.037610e-01,-4.687015e-01,-2.924961e-01,-4.526367e-01,3.924533e-01,5.317051e-01,...,-4.638313e-01,6.200173e-01,-4.785022e-01,9.271868e-01,-4.757013e-01,-2.308165e-01,-2.351254e-01,1.846324e-01,-5.022153e-01,7.751137e-01
max,1.403482e+00,1.719916e+00,1.600843e+00,1.437334e+00,2.609381e+00,2.669312e+00,4.972435e+00,3.148777e+00,2.429870e+00,2.743115e+00,...,3.072882e+00,2.935592e+00,2.667578e+00,1.813219e+00,2.670805e+00,6.438566e+00,5.911724e+00,1.563724e+00,2.214200e+00,2.558922e+00


In [ ]:
# View outcome varible split to determine if stratifying is warrented in train test split
print(y.value_counts())

class
1    112
3     72
2     61
5     52
4     49
6     20
Name: count, dtype: int64


In [ ]:
from sklearn.model_selection import train_test_split

# Train-test split - #1

X_train, X_test, y_train, y_test = train_test_split(
    X_scaled, y, test_size=0.2, random_state=42, stratify=y
)
# Stratify train test split to ensure all classes are balanced in split

In [ ]:
print(X_train.shape)
print(X_test.shape)
print(y_train.shape)
print(y_test.shape)

(292, 34)
(74, 34)
(292,)
(74,)


## Train Models - First Train_Test_Split

In [ ]:
# Import SVC Model - target data is categorical with labels 1-6
from sklearn.svm import SVC

In [ ]:
# Train first SVC model with a linear kernel and balanced class weight to compensate for class imbalance
svc_linear_1 = SVC(kernel='linear', C=1, class_weight='balanced')
svc_linear_1.fit(X_train, y_train)

SVC(C=1, class_weight='balanced', kernel='linear')

In [ ]:
# Train second SVC model with a polynomial kernel and balanced class weight to compensate for class imbalance
svc_poly_1 = SVC(kernel='poly', C=1.5, gamma=0.5, class_weight='balanced')
svc_poly_1.fit(X_train, y_train)

SVC(C=1.5, class_weight='balanced', gamma=0.5, kernel='poly')

In [ ]:
# Train third SVC model with a RBF kernel and balanced class weight to compensate for class imbalance
svc_rbf_1 = SVC(kernel='rbf', C=0.5, gamma='scale', class_weight='balanced')
svc_rbf_1.fit(X_train, y_train)

SVC(C=0.5, class_weight='balanced')

In [ ]:
# Import metrics libraries
from sklearn.metrics import (
    accuracy_score,
    classification_report,
    confusion_matrix,
)

In [ ]:
# Loop evalation metric print outs

# Define model name:object tuples
models = [
    ("Linear SVM 1", svc_linear_1),
    ("Polynomial SVM 1", svc_poly_1),
    ("RBF SVM 1", svc_rbf_1)
]
# Iterate loop for metrics per model
for name, model in models:
    print(f"\n=== {name} ===")
    y_pred = model.predict(X_test)

    # Accuracy
    print(f"Accuracy: {accuracy_score(y_test, y_pred):.4f}")

    # Classification Report
    print("\nClassification Report:")
    print(classification_report(y_test, y_pred))

    # Confusion Matrix
    print("Confusion Matrix:")
    print(confusion_matrix(y_test, y_pred))


=== Linear SVM 1 ===
Accuracy: 0.9730

Classification Report:
              precision    recall  f1-score   support

           1       1.00      1.00      1.00        23
           2       1.00      0.83      0.91        12
           3       1.00      1.00      1.00        15
           4       0.83      1.00      0.91        10
           5       1.00      1.00      1.00        10
           6       1.00      1.00      1.00         4

    accuracy                           0.97        74
   macro avg       0.97      0.97      0.97        74
weighted avg       0.98      0.97      0.97        74

Confusion Matrix:
[[23  0  0  0  0  0]
 [ 0 10  0  2  0  0]
 [ 0  0 15  0  0  0]
 [ 0  0  0 10  0  0]
 [ 0  0  0  0 10  0]
 [ 0  0  0  0  0  4]]

=== Polynomial SVM 1 ===
Accuracy: 0.9054

Classification Report:
              precision    recall  f1-score   support

           1       1.00      1.00      1.00        23
           2       1.00      0.67      0.80        12
           3       

The trained support vector machine (SVM) classification models used linear, polynomial, and RBF kernels on a stratified train-test split of a dermatology dataset. Stratification was applied to ensure that all six target classes were proportionally represented in both the training and testing data.

- The Linear SVM achieved the highest accuracy at 97.30%, correctly classifying nearly all of the 74 test instances. It attained perfect precision and recall (1.00) for classes 1, 3, 5, and 6. Class 2 had a recall of 0.83, with two of its instances misclassified as class 4 — as confirmed by the confusion matrix.

- The Polynomial SVM had a lower accuracy of 90.54%. While precision remained high across most classes, recall for class 2 dropped to 0.67 and for class 6 to 0.75. The confusion matrix shows that the polynomial model had difficulty distinguishing between classes 2 and 4, misclassifying four instances of class 2 as class 4, along with one instance each of classes 3, 5, and 6. These patterns suggest that the polynomial kernel may have overfit the training data or misrepresented certain feature interactions, leading to increased misclassification.

- The RBF SVM achieved an accuracy of 95.95%, performing comparably to the linear model. It demonstrated strong classification across all classes, with minor dips in recall for class 2 (0.83) and class 3 (0.93). Most predictions were correct, and misclassifications were minimal, as shown in the confusion matrix.

In [ ]:
# Perform a new Train-Test split

# Train-test split - #2

X_train2, X_test2, y_train2, y_test2 = train_test_split(
    X_scaled, y, test_size=0.2, random_state=99
)
# Removing the stratification to compare how the models perform with and without it

In [ ]:
print(X_train2.shape)
print(X_test2.shape)
print(y_train2.shape)
print(y_test2.shape)

(292, 34)
(74, 34)
(292,)
(74,)


## Train Models - Second Train_Test_Split (no stratification)

In [ ]:
# Train first SVC model with a linear kernel and balanced class weight to compensate for class imbalance (second train_test_split)
svc_linear_2 = SVC(kernel='linear', C=1, class_weight='balanced')
svc_linear_2.fit(X_train2, y_train2)

SVC(C=1, class_weight='balanced', kernel='linear')

In [ ]:
# Train second SVC model with a polynomial kernel and balanced class weight to compensate for class imbalance (second train_test_split)
svc_poly_2 = SVC(kernel='poly', C=1.5, gamma=0.5, class_weight='balanced')
svc_poly_2.fit(X_train2, y_train2)

SVC(C=1.5, class_weight='balanced', gamma=0.5, kernel='poly')

In [ ]:
# Train third SVC model with a RBF kernel and balanced class weight to compensate for class imbalance (second train_test_split)
svc_rbf_2 = SVC(kernel='rbf', C=0.5, gamma='scale', class_weight='balanced')
svc_rbf_2.fit(X_train2, y_train2)

SVC(C=0.5, class_weight='balanced')

In [ ]:
# Loop evalation metric print outs

# Define model name:object tuples
models2 = [
    ("Linear SVM 2", svc_linear_2),
    ("Polynomial SVM 2", svc_poly_2),
    ("RBF SVM 2", svc_rbf_2)
]
# Iterate loop for metrics per model
for name, model in models2:
    print(f"\n=== {name} ===")
    y_pred2 = model.predict(X_test2)

    # Accuracy
    print(f"Accuracy: {accuracy_score(y_test2, y_pred2):.4f}")

    # Classification Report
    print("\nClassification Report:")
    print(classification_report(y_test2, y_pred2))

    # Confusion Matrix
    print("Confusion Matrix:")
    print(confusion_matrix(y_test2, y_pred2))


=== Linear SVM 2 ===
Accuracy: 0.9865

Classification Report:
              precision    recall  f1-score   support

           1       1.00      1.00      1.00        23
           2       0.89      1.00      0.94         8
           3       1.00      1.00      1.00        18
           4       1.00      0.92      0.96        12
           5       1.00      1.00      1.00        13

    accuracy                           0.99        74
   macro avg       0.98      0.98      0.98        74
weighted avg       0.99      0.99      0.99        74

Confusion Matrix:
[[23  0  0  0  0]
 [ 0  8  0  0  0]
 [ 0  0 18  0  0]
 [ 0  1  0 11  0]
 [ 0  0  0  0 13]]

=== Polynomial SVM 2 ===
Accuracy: 0.9730

Classification Report:
              precision    recall  f1-score   support

           1       1.00      1.00      1.00        23
           2       0.88      0.88      0.88         8
           3       1.00      1.00      1.00        18
           4       0.92      0.92      0.92        12
 

In [ ]:
# The Confusion Matrices only showed 5x5, when there were 6 classes of target variables
# Checking the counts of y_test2 indicated that class 6 did not make it into the test set due to the removal of stratification
print(y_test2.value_counts())

class
1    23
3    18
5    13
4    12
2     8
Name: count, dtype: int64


In the second train-test split, there was no stratification in order to evaluate how the models would perform under naturally imbalanced class distributions. While all three SVM models performed slightly better on this split, the improvement is skewed because class 6 was not captured in the test set, which simplified the classification task and reduced opportunities for misclassification.

- The Linear SVM achieved 98.65% accuracy and nearly perfect performance. Precision and recall were 1.00 for most classes. Class 4 had a small drop in recall (0.92), and class 2’s precision dipped slightly to 0.89 due to one false positive.

- The Polynomial SVM achieved 97.30% accuracy. The model was strong overall, but slightly less reliable. Class 2 showed balanced but lower precision and recall (0.88 each), and class 4 also had a minor drop to 0.92 but misclassifications were minimal.

- The RBF SVM  achieved 98.65% accuracy and matched the Linear model almost exactly. Minor recall loss for class 4 (0.92), but otherwise showed excellent class separation. There were no major misclassifications observed.

These results demonstrate how excluding rare classes (like class 6) can boost metrics without truly improving model generalization. This reinforced the value of stratified sampling in performance evaluation and led me to do one additional train-test split with a different random_state and reintroducing stratification.

---

In [ ]:
# Train-test split - #3

X_train3, X_test3, y_train3, y_test3 = train_test_split(
    X_scaled, y, test_size=0.2, random_state=28, stratify=y
)

In [ ]:
print(X_train3.shape)
print(X_test3.shape)
print(y_train3.shape)
print(y_test3.shape)

(292, 34)
(74, 34)
(292,)
(74,)


In [ ]:
# Check immediately that all classes are represented
print(y_test3.value_counts())

class
1    23
3    15
2    12
5    10
4    10
6     4
Name: count, dtype: int64


In [ ]:
# Train first SVC model with a linear kernel and balanced class weight to compensate for class imbalance (third train_test_split)
svc_linear_3 = SVC(kernel='linear', C=1, class_weight='balanced')
svc_linear_3.fit(X_train3, y_train3)

SVC(C=1, class_weight='balanced', kernel='linear')

In [ ]:
# Train second SVC model with a polynomial kernel and balanced class weight to compensate for class imbalance (third train_test_split)
svc_poly_3 = SVC(kernel='poly', C=1.5, gamma=0.5, class_weight='balanced')
svc_poly_3.fit(X_train3, y_train3)

SVC(C=1.5, class_weight='balanced', gamma=0.5, kernel='poly')

In [ ]:
# Train third SVC model with a RBF kernel and balanced class weight to compensate for class imbalance (third train_test_split)
svc_rbf_3 = SVC(kernel='rbf', C=0.5, gamma='scale', class_weight='balanced')
svc_rbf_3.fit(X_train3, y_train3)

SVC(C=0.5, class_weight='balanced')

In [ ]:
# Loop evalation metric print outs

# Define model name:object tuples
models3 = [
    ("Linear SVM 3", svc_linear_3),
    ("Polynomial SVM 3", svc_poly_3),
    ("RBF SVM 3", svc_rbf_3)
]
# Iterate loop for metrics per model
for name, model in models3:
    print(f"\n=== {name} ===")
    y_pred3 = model.predict(X_test3)

    # Accuracy
    print(f"Accuracy: {accuracy_score(y_test3, y_pred3):.4f}")

    # Classification Report
    print("\nClassification Report:")
    print(classification_report(y_test3, y_pred3))

    # Confusion Matrix
    print("Confusion Matrix:")
    print(confusion_matrix(y_test3, y_pred3))


=== Linear SVM 3 ===
Accuracy: 0.9459

Classification Report:
              precision    recall  f1-score   support

           1       1.00      1.00      1.00        23
           2       0.83      0.83      0.83        12
           3       1.00      1.00      1.00        15
           4       0.80      0.80      0.80        10
           5       1.00      1.00      1.00        10
           6       1.00      1.00      1.00         4

    accuracy                           0.95        74
   macro avg       0.94      0.94      0.94        74
weighted avg       0.95      0.95      0.95        74

Confusion Matrix:
[[23  0  0  0  0  0]
 [ 0 10  0  2  0  0]
 [ 0  0 15  0  0  0]
 [ 0  2  0  8  0  0]
 [ 0  0  0  0 10  0]
 [ 0  0  0  0  0  4]]

=== Polynomial SVM 3 ===
Accuracy: 0.9189

Classification Report:
              precision    recall  f1-score   support

           1       1.00      1.00      1.00        23
           2       0.71      0.83      0.77        12
           3       

For the third train-test split, stratification was reintroduced to ensure a representative distribution of all 6 target classes in both the training and testing data. This approach aimed to validate earlier results and confirm that each model could generalize well across the full set of classes, including the low-frequency class 6.

- The Linear SVM achieved an accuracy of 94.59% and maintained strong performance overall. Precision and recall were perfect for most classes, but recall for class 2 and class 4 dropped to 0.83 and 0.80, respectively. The confusion matrix shows two instances of class 4 misclassified as class 2 and vice versa, indicating some overlap between these two classes.

- The Polynomial SVM reached an accuracy of 91.89%, which is consistent with its earlier performance. Class 2 had a moderate recall of 0.83, while class 4 dropped further to 0.70. Multiple misclassifications suggest that the polynomial kernel may be overfitting or struggling to separate class boundaries.

- The RBF SVM achieved an accuracy of 95.95%, matching its result from the first stratified split and outperforming the other two models on this split. It showed perfect precision and recall for 4 out of 6 classes, with only minor dips in recall for class 2 (0.83) and class 4 (0.90). The confusion matrix reveals just a few misclassifications, again centered on classes 2 and 4, indicating otherwise solid performance.

These results reaffirm that stratification improves the reliability of model evaluation. It ensures that all classes, especially underrepresented ones like class 6, are properly included in the test set, and it helps uncover areas of confusion between similar classes, as seen between classes 2 and 4.

---

Identify Top 10% Predictive Features

Using 'svc_linear_1' model, the top 10% the most predictive features using the .coef_ attribute will be selected. The linear SVMwas used because only linear models produce directly interpretable feature weights via .coef_. Polynomial and RBF kernels project the data into higher-dimensional spaces, where feature weights cannot be directly interpreted or translated into individual relationships.

In [ ]:
# Get the model coefficients
coefs = svc_linear_1.coef_
coefs

array([[ 5.05252657e-02,  1.02847357e-01,  1.30028511e-01,
         4.61270089e-03,  9.26513531e-02,  0.00000000e+00,
        -1.38777878e-17, -2.77555756e-17,  2.33316639e-01,
         3.99223049e-02,  9.35432867e-03,  0.00000000e+00,
        -7.44402045e-02,  2.66164388e-02, -1.38777878e-17,
        -1.30323843e-01,  4.08267874e-02,  4.31284757e-02,
         4.85132096e-02,  2.32036666e-01,  1.60516499e-01,
         2.75321798e-01, -7.82705280e-02,  7.93870967e-02,
         0.00000000e+00,  4.82875147e-02, -2.77555756e-17,
        -5.06360449e-01,  0.00000000e+00, -1.38777878e-17,
         0.00000000e+00,  4.86567121e-02, -2.77555756e-17,
         8.26734006e-02],
       [ 1.26300756e-02,  4.65530047e-02, -1.08765264e-02,
        -3.39053924e-02, -2.75453312e-03, -1.03680121e-01,
         0.00000000e+00, -8.54809031e-02,  5.42132389e-02,
         1.59110504e-02,  0.00000000e+00, -9.47223692e-02,
        -1.62911722e-02,  5.29322322e-02,  0.00000000e+00,
        -3.70037809e-02, -2.32

In [ ]:
# Convert to absolute values of coefs
import numpy as np
abs_coefs = np.abs(coefs)

In [ ]:
# Calculate the average imporatance across all classes
avg_importance = np.mean(abs_coefs, axis=0)

In [ ]:
# Extract the top 10% of features
num_top = int(0.10 * len(avg_importance))  # 10% of 34 = 3.4 → becomes 3
top_indices = np.argsort(avg_importance)[-num_top:]  # sorts and extracts the top 3

In [ ]:
# View top 3 as their original column names
top_features = X.columns[top_indices]
print("Top predictive features:", top_features.tolist())

Top predictive features: ['koebner phenomenon', 'disappearance of the granular layer', 'spongiosis']


In [ ]:
# Extract the top features into a new dataframe
X_top = X_scaled[top_features]
X_top.head()

,koebner phenomenon,disappearance of the granular layer,spongiosis
0,-0.699049,-0.537770,1.813219
1,0.403761,-0.537770,-0.844878
2,0.403761,-0.537770,1.813219
3,-0.699049,2.935592,-0.844878
4,1.506571,1.777805,0.927187


In [ ]:
# Peform a new Train_Test_Split
X_train_top, X_test_top, y_train_top, y_test_top = train_test_split(
    X_top, y, test_size=0.2, random_state=33, stratify=y
)

In [ ]:
# Train first SVC model with a linear kernel and balanced class weight to compensate for class imbalance (Top 10% train_test_split)
svc_linear_top = SVC(kernel='linear', C=1, class_weight='balanced')
svc_linear_top.fit(X_train_top, y_train_top)

SVC(C=1, class_weight='balanced', kernel='linear')

In [ ]:
# Train second SVC model with a polynomial kernel and balanced class weight to compensate for class imbalance (Top 10% train_test_split)
svc_poly_top = SVC(kernel='poly', C=1.5, gamma=0.5, class_weight='balanced')
svc_poly_top.fit(X_train_top, y_train_top)

SVC(C=1.5, class_weight='balanced', gamma=0.5, kernel='poly')

In [ ]:
# Train third SVC model with a RBF kernel and balanced class weight to compensate for class imbalance (Top 10% train_test_split)
svc_rbf_top = SVC(kernel='rbf', C=0.5, gamma='scale', class_weight='balanced')
svc_rbf_top.fit(X_train_top, y_train_top)

SVC(C=0.5, class_weight='balanced')

In [ ]:
# Loop evalation metric print outs

# Define model name:object tuples
models_top = [
    ("Linear SVM Top 10%", svc_linear_top),
    ("Polynomial SVM Top 10%", svc_poly_top),
    ("RBF SVM Top 10%", svc_rbf_top)
]
# Iterate loop for metrics per model
for name, model in models_top:
    print(f"\n=== {name} ===")
    y_pred_top = model.predict(X_test_top)

    # Accuracy
    print(f"Accuracy: {accuracy_score(y_test_top, y_pred_top):.4f}")

    # Classification Report
    print("\nClassification Report:")
    print(classification_report(y_test_top, y_pred_top))

    # Confusion Matrix
    print("Confusion Matrix:")
    print(confusion_matrix(y_test_top, y_pred_top))


=== Linear SVM Top 10% ===
Accuracy: 0.6622

Classification Report:
              precision    recall  f1-score   support

           1       1.00      0.48      0.65        23
           2       0.69      0.92      0.79        12
           3       0.41      0.47      0.44        15
           4       0.71      1.00      0.83        10
           5       0.62      1.00      0.77        10
           6       0.00      0.00      0.00         4

    accuracy                           0.66        74
   macro avg       0.57      0.64      0.58        74
weighted avg       0.69      0.66      0.63        74

Confusion Matrix:
[[11  0 10  0  2  0]
 [ 0 11  0  0  1  0]
 [ 0  2  7  4  2  0]
 [ 0  0  0 10  0  0]
 [ 0  0  0  0 10  0]
 [ 0  3  0  0  1  0]]

=== Polynomial SVM Top 10% ===
Accuracy: 0.6351

Classification Report:
              precision    recall  f1-score   support

           1       0.92      0.48      0.63        23
           2       0.69      0.92      0.79        12
       

/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


For the final set of models, only the top 10% most predictive features were used, identified from the linear SVM's learned coefficients, to assess how well the models could perform with a reduced feature set. This step aimed to test whether a simpler, more interpretable model could still achieve competitive accuracy. While feature selection can reduce overfitting and improve model explainability, the results here showed a significant drop in performance.

- The Linear SVM had an accuracy score of 66.22%. Performance declined sharply compared to the full feature model. Class 1 recall dropped to 0.48 despite perfect precision, meaning over half of class 1 samples were misclassified. Class 6 wasn't predicted at all, and class 3 showed both weak recall and precision.

- The Polynomial SVM reached an accuracy of 63.51%, slightly lower than the linear model. It struggled with class 1 and class 3 in particular, and class 6 had both low recall (0.25) and precision (0.33). Misclassifications were spread across multiple classes, suggesting overfitting and poor generalization.

- The RBF SVM had an accuracy score of 63.51% and mirrored the polynomial model's metrics and confusion matrix exactly. This is likely due to the limited features available to train on and the similarities between the polynomial and RBF models.

Overall, these results demonstrate that reducing the feature set to only the top 10% significantly weakened the models' ability to distinguish between all 6 classes. While feature selection is valuable in some contexts, this task requires more than 10% of features to provide reliable performance.

---

## Conclusion
This investigation applied supervised machine learning techniques to a dermatology dataset for multiclass classification. The dataset included 34 features and 6 distinct diagnostic classes. Initial preprocessing involved filling missing values, standardizing all features using StandardScaler, and ensuring the target variable was in a compatible format. Three Support Vector Machine (SVM) classifiers, Linear, Polynomial, and RBF, were trained and evaluated across three different train-test splits. The first split used stratification to maintain class balance; the second intentionally omitted stratification to examine the impact of uneven class distribution; and the third reintroduced stratification to validate model performance under more controlled conditions. Across these nine models, the Linear and RBF SVMs consistently outperformed the Polynomial SVM, with the highest accuracy in the unstratified split reaching 98.65% with the RBF model; and 97.30% in the stratified linear model.

Additonally, the top 10% most predictive features based on the linear model's coefficients were selected and used to retrain all three models. These reduced-feature models showed significantly lower performance (around 63-66% accuracy), especially struggling with underrepresented classes like class 6.

Overall, the results demonstrate that full-feature models offer improved generalization compared to reduced-feature models. Additionally, though the unstratified models achieved slightly higher accuracy scores, those results were misleading due to the absence of rare classes like class 6 in the test set. The Linear SVM, trained on the first stratified split, provided the most reliable and balanced performance across all six classes, making it the best overall model in terms of generalization and fairness.